README.rst
sublime-info
============

.. image:: https://travis-ci.org/twolfson/sublime-info.png?branch=master
   :target: https://travis-ci.org/twolfson/sublime-info
   :alt: Build Status

Gather information about `Sublime Text`_

This was built for the `Sublime plugin tests`_ framework. It provides a cross-platform way to collect data about `Sublime Text`_ without running `Sublime Text`_ (necessary for creating a plugin harness to `Sublime Text`_).

.. _`Sublime Text`: http://sublimetext.com/
.. _`Sublime plugin tests`: https://github.com/twolfson/sublime-plugin-tests

    Currently, only Linux is supported but OSX and Windows support are planned.

Getting Started
---------------
Install the module with: ``pip install sublime_info``

.. code:: python

    import sublime_info
    sublime_info.get_sublime_path()  # /usr/bin/subl
    sublime_info.get_sublime_version()  # 3083
    sublime_info.get_package_directory()  # /home/todd/.config/sublime-text-2/Packages

Documentation
-------------
``sublime_info`` provides 3 top level functions for your consumption.

get_sublime_path
^^^^^^^^^^^^^^^^
.. code:: python

    sublime_info.get_sublime_path()
    """Resolve Sublime Text path (e.g. /usr/bin/subl)

    If ``SUBLIME_TEXT_PATH`` is provided via environment variables, it will be used.
    Otherwise, a ``which``-like resolution will be returned.

    :raises STNotFoundError: If Sublime Text cannot be found, an error will be raised.
    :returns: ``SUBLIME_TEXT_PATH`` or ``which``-like resolution
    :rtype: str
    """

get_sublime_version
^^^^^^^^^^^^^^^^^^^
.. code:: python

    sublime_info.get_sublime_version()
    """Resolve Sublime Text version (e.g. 2221, 3083)

    Sublime Text is resolved via ``get_sublime_path``

    :raises Exception: If the Sublime Text version cannot be parsed, an error will be raised.
    :returns: Version of Sublime Text returned by ``sublime_text --version``.
    :rtype: int
    """

get_package_directory
^^^^^^^^^^^^^^^^^^^^^
.. code:: python

    sublime_info.get_package_directory()
    """Resolve Sublime Text package directory (e.g. /home/todd/.config/sublime-text-2/Packages)

    :raises Exception: If the Sublime Text version is not recognized, an error will be raised.
    :returns: Path to Sublime Text's package directory
    :rtype: str
    """

Contributing
------------
In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Test via ``./test_linux.sh``.

Donating
--------
Support this project and `others by twolfson`_ via `gittip`_.

.. image:: https://rawgithub.com/twolfson/gittip-badge/master/dist/gittip.png
   :target: `gittip`_
   :alt: Support via Gittip

.. _`others by twolfson`:
.. _gittip: https://www.gittip.com/twolfson/

Unlicense
---------
As of Oct 14 2013, Todd Wolfson has released this repository and its contents to the public domain.

It has been released under the `UNLICENSE`_.

.. _UNLICENSE: https://github.com/twolfson/sublime-info/blob/master/UNLICENSE


CHANGELOG.rst
sublime-info changelog
======================
0.2.0 - Added Python 3 support via @adambullmer in #5. Fixes #4

0.1.3 - Repaired Travis CI

0.1.2 - Added foundry for release

0.1.1 - Fix for README in PyPI

0.1.0 - Initial release

0.0.3 - Added support for package directory

0.0.2 - Added support for build number

0.0.1 - Resolving Sublime in multiple forms


test.sh
#!/usr/bin/env bash
# Exit on first error
set -e

# Run our tests
nosetests --nocapture


test_linux.sh
#!/usr/bin/env bash
# Exit on first error
set -e

# Run our tests
EXPECTED_PATH="/usr/bin/sublime_text" EXPECTED_VERSION="2221" EXPECTED_PKG_DIR="/home/todd/.config/sublime-text-2/Packages" nosetests --nocapture
SUBLIME_TEXT_PATH="null" EXPECT_ERROR="1" nosetests --nocapture


test/install.sh
#!/usr/bin/env bash
# Exit on first error
set -e

# If there is a Sublime Text version, then install Sublime
if test -n "$SUBLIME_TEXT_VERSION"; then
  # Use script for ease of use
  curl --location https://rawgit.com/twolfson/sublime-installer/0.2.2/install.sh | sh -s "$SUBLIME_TEXT_VERSION"

  # Output Sublime version
  subl --version

  # If there is a rename, rename subl
  if test -n "$SUBLIME_TEXT_RENAME"; then
    sudo mv /usr/bin/subl "/usr/bin/$SUBLIME_TEXT_RENAME"
  fi
fi


sublime_info/sublime_info.py
# Load in core, 3rd party, and local dependencies
import os
import re
import subprocess
from shutilwhich import which
from errors import STNotResolvedError, STBadLocationError


class SublimeInfo(object):
    # Load in an environment variable constant
    sublime_path=os.environ.get('SUBLIME_TEXT_PATH', None)

    # Define init
    def __init__(self, sublime_path=None):
        # Allow for customizaton of sublime_path
        if sublime_path:
            self.sublime_path = sublime_path

    # Define internal lookup which ignores
    @classmethod
    def _get_sublime_path(cls):
        # Attempt to resolve Sublime Text
        path = (which('subl') or
                which('sublime_text'))

        # If Sublime is not found, raise our exception
        if not path:
            raise STNotResolvedError(
                    'Sublime Text could not be found via the command "%s" or "%s"' %
                    ('subl',
                     'sublime_text'))

        # Otherwise, return the path
        return path


    @classmethod
    def get_sublime_path(cls):
        """Resolve Sublime Text path (e.g. /usr/bin/subl)

        If ``SUBLIME_TEXT_PATH`` is provided via environment variables, it will be used.
        Otherwise, a ``which``-like resolution will be returned.

        :raises STNotFoundError: If Sublime Text cannot be found, an error will be raised.
        :returns: ``SUBLIME_TEXT_PATH`` or ``which``-like resolution
        :rtype: str
        """
        # If sublime_path is provided, verify it exists
        sublime_path = cls.sublime_path
        if sublime_path:
            if not os.path.exists(sublime_path):
                raise STBadLocationError(
                        'Sublime Text could not be found at "%s"' % sublime_path)
        # Otherwise, use the internal lookup
        else:
            sublime_path = cls._get_sublime_path()

        # Return the found path
        return sublime_path

    @classmethod
    def get_sublime_version(cls):
        """Resolve Sublime Text version (e.g. 2221, 3083)

        Sublime Text is resolved via ``get_sublime_path``

        :raises Exception: If the Sublime Text version cannot be parsed, an error will be raised.
        :returns: Version of Sublime Text returned by ``sublime_text --version``.
        :rtype: int
        """
        # Get the path to sublime and grab the version
        sublime_path = cls.get_sublime_path()
        child = subprocess.Popen([sublime_path, '--version'], stdout=subprocess.PIPE)
        version_stdout = str(child.stdout.read())

        # Kill the child
        child.kill()

        # Parse out build number from stdout
        # Sublime Text 2 Build 2221
        # Sublime Text Build 3083
        version_match = re.search(r'\d{4}', version_stdout)
        if not version_match:
            raise Exception('Sublime Text version not found in "%s"' % version_stdout)

        # Coerce and return the version
        return int(version_match.group(0))

    @classmethod
    def get_package_directory(cls):
        """Resolve Sublime Text package directory (e.g. /home/todd/.config/sublime-text-2/Packages)

        :raises Exception: If the Sublime Text version is not recognized, an error will be raised.
        :returns: Path to Sublime Text's package directory
        :rtype: str
        """
        # TODO: On Windows, OSX these will not be the same
        # Get the version
        version = cls.get_sublime_version()

        # Run Linux-only logic for pkg_dir
        pkg_dir = None
        if version >= 2000 and version < 3000:
            pkg_dir = os.path.expanduser('~/.config/sublime-text-2/Packages')
        elif version >= 3000 and version < 4000:
            pkg_dir = os.path.expanduser('~/.config/sublime-text-3/Packages')

        # Assert the package dir was found
        if not pkg_dir:
            raise Exception('Sublime Text version "%s" not recognized' % version)

        # Return the package directory
        return pkg_dir


sublime_info/SublimeInfo.py
# Load in core, 3rd party, and local dependencies
from __future__ import absolute_import
import os
import re
import subprocess
from sublime_info.errors import STNotResolvedError, STBadLocationError
try:
    from shutil import which
except ImportError:  # Python 2 fallback
    from shutilwhich import which


class SublimeInfo(object):
    # Load in an environment variable constant
    sublime_path=os.environ.get('SUBLIME_TEXT_PATH', None)

    # Define init
    def __init__(self, sublime_path=None):
        # Allow for customizaton of sublime_path
        if sublime_path:
            self.sublime_path = sublime_path

    # Define internal lookup which ignores
    @classmethod
    def _get_sublime_path(cls):
        # Attempt to resolve Sublime Text
        path = (which('subl') or
                which('sublime_text'))

        # If Sublime is not found, raise our exception
        if not path:
            raise STNotResolvedError(
                    'Sublime Text could not be found via the command "%s" or "%s"' %
                    ('subl',
                     'sublime_text'))

        # Otherwise, return the path
        return path


    @classmethod
    def get_sublime_path(cls):
        """Resolve Sublime Text path (e.g. /usr/bin/subl)

        If ``SUBLIME_TEXT_PATH`` is provided via environment variables, it will be used.
        Otherwise, a ``which``-like resolution will be returned.

        :raises STNotFoundError: If Sublime Text cannot be found, an error will be raised.
        :returns: ``SUBLIME_TEXT_PATH`` or ``which``-like resolution
        :rtype: str
        """
        # If sublime_path is provided, verify it exists
        sublime_path = cls.sublime_path
        if sublime_path:
            if not os.path.exists(sublime_path):
                raise STBadLocationError(
                        'Sublime Text could not be found at "%s"' % sublime_path)
        # Otherwise, use the internal lookup
        else:
            sublime_path = cls._get_sublime_path()

        # Return the found path
        return sublime_path

    @classmethod
    def get_sublime_version(cls):
        """Resolve Sublime Text version (e.g. 2221, 3083)

        Sublime Text is resolved via ``get_sublime_path``

        :raises Exception: If the Sublime Text version cannot be parsed, an error will be raised.
        :returns: Version of Sublime Text returned by ``sublime_text --version``.
        :rtype: int
        """
        # Get the path to sublime and grab the version
        sublime_path = cls.get_sublime_path()
        child = subprocess.Popen([sublime_path, '--version'], stdout=subprocess.PIPE)
        version_stdout = str(child.stdout.read())

        # Kill the child
        child.kill()

        # Parse out build number from stdout
        # Sublime Text 2 Build 2221
        # Sublime Text Build 3083
        version_match = re.search(r'\d{4}', version_stdout)
        if not version_match:
            raise Exception('Sublime Text version not found in "%s"' % version_stdout)

        # Coerce and return the version
        return int(version_match.group(0))

    @classmethod
    def get_package_directory(cls):
        """Resolve Sublime Text package directory (e.g. /home/todd/.config/sublime-text-2/Packages)

        :raises Exception: If the Sublime Text version is not recognized, an error will be raised.
        :returns: Path to Sublime Text's package directory
        :rtype: str
        """
        # TODO: On Windows, OSX these will not be the same
        # Get the version
        version = cls.get_sublime_version()

        # Run Linux-only logic for pkg_dir
        pkg_dir = None
        if version >= 2000 and version < 3000:
            pkg_dir = os.path.expanduser('~/.config/sublime-text-2/Packages')
        elif version >= 3000 and version < 4000:
            pkg_dir = os.path.expanduser('~/.config/sublime-text-3/Packages')

        # Assert the package dir was found
        if not pkg_dir:
            raise Exception('Sublime Text version "%s" not recognized' % version)

        # Return the package directory
        return pkg_dir
